### PR TITLE
Step 11 - Tripal Importer - Chado API requirements/modifications

### PR DESCRIPTION
**Pull request by risharde**
MVIEWS API added, CV API added, modifications to the query api, minor format changes required for Tripal Importers to work
@laceysanderson

### CODE REVIEW

**Comment by laceysanderson** [comment]
We need to remove the allowRowCount() from the chado_query() function that you added in this PR. Counting using the row count is only valid if this is an update, delete, insert and we can't ensure that's the query type being used in chado_query. That's why it's deprecated now because it is misleading in select queries.

For the importers, you will want to use db_insert/update which will allow row count or db_select once we switch to the new biodb API... What are you using the counts for? An alternative is just to through a count specific query in there (i.e. SELECT count(*) FROM chado.table).

**Comment by risharde** [comment]
Yikes, this is definitely going to break the loaders. I'll try to track back where rowCounts are needed, I think it might be in a different PR that's already been approved

**Comment by risharde** [comment]
@laceysanderson I removed the rowCounts but I'm pretty sure as you said, it will need to be update for the loaders. Trying to track this back. I'm hoping it won't be crazy hard to resolve.

**Comment by risharde** [comment]
@laceysanderson I made the rowCount changes to the importer reviewed branch but I didn't create a PR for this branch since I think we're compiling everything in there before we do the PR to merge to main right?

**Comment by laceysanderson** [comment]
> @laceysanderson I made the rowCount changes to the importer reviewed branch but I didn't create a PR for this branch since I think we're compiling everything in there before we do the PR to merge to main right?

@risharde you should have made a PR for that. The tripal importer branch should only have code that has been reviewed by others. Then we will make a PR for the tripal importer reviewed branch where we will check functionality. In the future please do not make and edits directly to the tripal importer reviewed branch. For this case, please comment here with the commits where the changes were made so I can review them.

File: tripal_chado/src/api/tripal_chado.query.api.inc
Comment (laceysanderson) [inline tripal_chado/src/api/tripal_chado.query.api.inc]: There should never be spaces between the function and it's docblock. Please remove these.

File: tripal_chado/src/api/tripal_chado.query.api.inc
Comment (laceysanderson) [inline tripal_chado/src/api/tripal_chado.query.api.inc]: Remove these commented out versions. It makes the code quite messy.

File: tripal_chado/src/api/tripal_chado.query.api.inc
Comment (laceysanderson) [inline tripal_chado/src/api/tripal_chado.query.api.inc]: Please remove debugging code. This branch will be merged into 9.x-4.x since it is not importer specific and we want to minimize divergence so we need to clean it up :-)

File: tripal_chado/src/api/tripal_chado.query.api.inc
Comment (laceysanderson) [inline tripal_chado/src/api/tripal_chado.query.api.inc]: Again, make sure to remove all debugging code, code-based comments and unnecessary white-space from this PR.

File: tripal_chado/src/api/tripal_chado.query.api.inc
Comment (laceysanderson) [inline tripal_chado/src/api/tripal_chado.query.api.inc]: These need to be full-on removed as this is not the Drupal approach.

**Comment by risharde** [comment]
@laceysanderson I don't know how to do that when you delete the original PR branches, maybe you can give me an idea, thanks.

**Review by laceysanderson — CHANGES_REQUESTED** [review]
There need to be automated tests for these (even just simple ones that call the function) to ensure that this code is run for deprecation notices.

**Comment by laceysanderson** [comment]
Now that I've done a more thorough review, I think I need to clarify your above comment on rowCount. 

Did you make changes to the tv4-tripalImporterReviewed branch? 

If so, you need to comment back here with links to those commits. You can do that by [going to the commit list for that branch -click on the branch under branches and then commits](https://github.com/tripal/t4d8/commits/tv4-tripalImporterReviewed), find the specific commits in that list and copy the commit hash (shortcode in blue at the right side of each commit in that listing) into a comment here.

**Comment by risharde** [comment]
Thanks @laceysanderson , sorry about this, I didn't think about just branching out from review and then making a PR. That might have been a better approach. 

These are the commits - it was small changes which hopefully you can allow. These basically remove the rowCount and instead does counts by doing separate SQL statements using COUNT(*)

https://github.com/tripal/t4d8/commit/07c4ad253ceee6f2fa4d2d7dd1a78bda43bb020e
https://github.com/tripal/t4d8/commit/0929c33d105ffd1017527bd40cbb2fe4b2803fe7

File: tripal_chado/src/api/tripal_chado.query.api.inc
Comment (risharde) [reply tripal_chado/src/api/tripal_chado.query.api.inc]: Thanks, I worked on this and it will be available in the next commit.

File: tripal_chado/src/api/tripal_chado.query.api.inc
Comment (risharde) [reply tripal_chado/src/api/tripal_chado.query.api.inc]: Removed all of these comments out which will be reflected in the next commit, thanks

File: tripal_chado/src/api/tripal_chado.query.api.inc
Comment (risharde) [reply tripal_chado/src/api/tripal_chado.query.api.inc]: Removed these comments as well, thanks! Will be seen in the upcoming commit.

File: tripal_chado/src/api/tripal_chado.query.api.inc
Comment (risharde) [reply tripal_chado/src/api/tripal_chado.query.api.inc]: Removed this one, luckily I did not do too much of these! Thanks!

File: tripal_chado/src/api/tripal_chado.query.api.inc
Comment (risharde) [reply tripal_chado/src/api/tripal_chado.query.api.inc]: These have also been removed, thanks for a thorough review!

**Comment by risharde** [comment]
@laceysanderson Just to confirm the tests you want me to write, I see we seem to already have tests for CV and QUERY. So I'm guessing you want for the MVIEWS? If not, could you correct me on what needs to be done in a little more detail. Thanks much!

**Comment by laceysanderson** [comment]
Functions to make tests for:

- CV: chado_insert_obo(), chado_get_obo(), chado_update_cvtermpath(), chado_clear_cvtermpath(), tripal_set_default_cv()
- MAPPING: tripal_chado_map_cvterms(), tripal_chado_add_cvterm_mapping()
- MVIEWS: chado_add_mview(), chado_edit_mview(), chado_get_mview_id(), chado_get_mview_table_names(), chado_refresh_mview(), chado_get_mviews(), chado_delete_mview(), chado_populate_mview().
- None for Chado Query API

Note: At a minimum you want to call the function and check it didn't fail. You should group multiple function calls into a single test method as it makes logical sense to do so (i.e. as the functions are related and would likely be tested together in the future).

**Comment by laceysanderson** [comment]
> Thanks @laceysanderson , sorry about this, I didn't think about just branching out from review and then making a PR. That might have been a better approach.
> 
> These are the commits - it was small changes which hopefully you can allow. These basically remove the rowCount and instead does counts by doing separate SQL statements using COUNT(*)
> 
> [07c4ad2](https://github.com/tripal/t4d8/commit/07c4ad253ceee6f2fa4d2d7dd1a78bda43bb020e) [0929c33](https://github.com/tripal/t4d8/commit/0929c33d105ffd1017527bd40cbb2fe4b2803fe7)

Yes, in the future definitely use the branch method you described here.

I checked these two commits and consider it ok. I did add an additional commit removing some unnecessary comments to clean up the code base.

**Comment by risharde** [comment]
@laceysanderson I started writing a test function to call the functions you specified with logic inbetween to assert etc. Little did I realize until after coding about 40 lines or so that these functions are erroring out. Please see this example below:

Drupal\Core\Database\DatabaseExceptionWrapper: SQLSTATE[42P01]: Undefined table: 7 ERROR:  relation "test41295788tripal_cv_obo" does not exist

This happens when I try to call chado_insert_obo("test","test") and it seems like it has something to do with the database not being initialized with this table.

I looked back at ChadoDbAPITest to discover that those tests are also marked incomplete for I guess the same reason?
Should I continue this exercise of adding all of these functions? Technically I feel like I'm coding in the dark if they will just error out similarly?

**Comment by risharde** [comment]
@laceysanderson following up on this issue, @spficklin looked at it while in meeting today as well and it's probably because the create table functions for this table not in this current branch (it seems to be here: https://github.com/tripal/t4d8/pull/160/files). I already just asked him if there's a way to somehow test what's in that branch on this current branch I'm working on for the unit tests but I'm also throwing this question at you as well just in case you have time  to think about it and IF there's a solution - I'm a bit mind blocked on how to proceed - I don't think writing blind test code would be a good approach but I am subjected by your guidance as well.

**Comment by laceysanderson** [comment]
Oh darn. I had missed that connection. In that case, this will have to go into the importers reviewed branch as well and we can skip writing the tests for now.

**Review by laceysanderson — APPROVED** [review]
I've switched the base to the importer reviewed branch and updated this branch so that the required tests can run. When these tests pass we can merge it;

**Pull request merged**
